### PR TITLE
feat(vlang): create module

### PR DESCRIPTION
**Pull request by Milo123459**
#### Description

Adds a module to display the currently installed V  version
#### Motivation and Context

Closes #

#### Screenshots (if appropriate):
img
#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [x] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: docs/config/README.md [hunk 1]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
 format = "via [⍱ $version](bold white) "
 ```
 
+## V
+
+The `v` module shows you your currently installed version of V.
+By default the module will be shown if any of the following conditions are met:
+- The current directory contains a file with the extension of .v
```
Comment (vladimyr) [inline docs/config/README.md:2735]:
```suggestion
- The current directory contains a file with the extension of `.v`
```

File: docs/config/README.md [hunk 2]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
...
+
+| Option              | Default                                           | Description                                     |
+| ------------------- | ------------------------------------------------- | ----------------------------------------------- |
+| `format`            | `"via [$symbol($version )]($style)"`              | The format for the module.                      |
+| `symbol`            | `"V "`                                           | A format string representing the symbol of V |
+| `detect_extensions` | `["v"]`                                        | Which extensions should trigger this module.    |
+| `detect_files`      | `[]` | Which filenames should trigger this module.     |
+| `detect_folders`    | `[]`                                  | Which folders should trigger this module.       |
+| `style`             | `"blue bold"`                                     | The style for the module.                       |
+| `disabled`          | `false`                                           | Disables the `deno` module.                     |
```
Comment (vladimyr) [inline docs/config/README.md:2747]:
```suggestion
| `disabled`          | `false`                                           | Disables the `vlang` module.                    |
```
:smiley:

File: docs/config/README.md [hunk 3]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
...
+
+### Variables
+
+| Variable | Example  | Description                          |
+| -------- | -------- | ------------------------------------ |
+| version  | `v0.2` | The version of `v`                |
```
Comment (vladimyr) [inline docs/config/README.md:2753]:
```suggestion
| version  | `v0.2`   | The version of `v`                  |
```

File: docs/config/README.md [hunk 4]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
...
+| symbol   |          | Mirrors the value of option `symbol` |
+| style\*  |          | Mirrors the value of option `style`  |
+
+### Example
+
+```toml
+# ~/.config/starship.toml
+[v]
+format = "via [V $version](bold bold) "
```
Comment (vladimyr) [inline docs/config/README.md:2762]: Wait, what? `bold` x2? :confused:

File: src/configs/starship_root.rs [hunk 5]
```diff
@@ -56,6 +56,7 @@ pub const PROMPT_ORDER: &[&str] = &[
     "scala",
     "swift",
     "terraform",
+    "v",
```
Comment (vladimyr) [inline src/configs/starship_root.rs:59]: If `go` is marked as `golang` this should be `vlang` instead.
```suggestion
    "vlang",
```

File: src/configs/mod.rs [hunk 6]
```diff
@@ -124,6 +125,7 @@ pub struct FullConfig<'a> {
     terraform: terraform::TerraformConfig<'a>,
     time: time::TimeConfig<'a>,
     username: username::UsernameConfig<'a>,
+    v: v::VConfig<'a>,
```
Comment (vladimyr) [inline src/configs/mod.rs:128]:
```suggestion
    vlang: v::VConfig<'a>,
```

File: docs/config/README.md [hunk 7]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
 format = "via [⍱ $version](bold white) "
 ```
 
+## V
```
Comment (vladimyr) [inline docs/config/README.md:2731]:
```suggestion
## Vlang
```

File: docs/config/README.md [hunk 8]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
...
+
+The `v` module shows you your currently installed version of V.
```
Comment (vladimyr) [inline docs/config/README.md:2733]:
```suggestion
The `vlang` module shows you your currently installed version of V.
```

File: docs/config/README.md [hunk 9]
```diff
@@ -220,6 +220,7 @@ $rust\
 $scala\
 $swift\
 $terraform\
+$v\
```
Comment (vladimyr) [inline docs/config/README.md:223]:
```suggestion
$vlang\
```

File: src/modules/mod.rs [hunk 10]
```diff
@@ -126,6 +127,7 @@ pub fn handle<'a>(module: &str, context: &'a Context) -> Option<Module<'a>> {
             "time" => time::module(context),
             "crystal" => crystal::module(context),
             "username" => username::module(context),
+            "v" => v::module(context),
```
Comment (vladimyr) [inline src/modules/mod.rs:130]:
```suggestion
            "vlang" => vlang::module(context),
```

File: src/modules/mod.rs [hunk 11]
```diff
@@ -53,6 +53,7 @@ mod terraform;
 mod time;
 mod username;
 mod utils;
+mod v;
```
Comment (vladimyr) [inline src/modules/mod.rs:56]:
```suggestion
mod vlang;
```

File: src/modules/v.rs [hunk 12]
```diff
@@ -0,0 +1,92 @@
+use super::{Context, Module, RootModuleConfig};
```
Comment (vladimyr) [inline src/modules/v.rs:1]: Rename module from `src/modules/v.rs` to `src/modules/vlang.rs`.

File: src/modules/v.rs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+use crate::configs::v::VConfig;
+use crate::formatter::StringFormatter;
+
+/// Creates a module with the current V version
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("v");
```
Comment (vladimyr) [inline src/modules/v.rs:8]:
```suggestion
    let mut module = context.new_module("vlang");
```

File: src/modules/v.rs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    #[test]
+    fn test_parse_v_version() {
+        const OUTPUT: &str = "V 0.2 30c0659\n";
+        assert_eq!(parse_v_version(OUTPUT.trim()), Some("v0.2".to_string()))
+    }
+
+    #[test]
+    fn folder_without_v_files() -> io::Result<()> {
+        let dir = tempfile::tempdir()?;
+        let actual = ModuleRenderer::new("v").path(dir.path()).collect();
```
Comment (vladimyr) [inline src/modules/v.rs:77]:
```suggestion
        let actual = ModuleRenderer::new("vlang").path(dir.path()).collect();
```

File: src/modules/v.rs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        let expected = None;
+        assert_eq!(expected, actual);
+        dir.close()
+    }
+
+    #[test]
+    fn folder_with_v_files() -> io::Result<()> {
+        let dir = tempfile::tempdir()?;
+        File::create(dir.path().join("hello.v"))?.sync_all()?;
+        let actual = ModuleRenderer::new("v").path(dir.path()).collect();
```
Comment (vladimyr) [inline src/modules/v.rs:87]:
```suggestion
        let actual = ModuleRenderer::new("vlang").path(dir.path()).collect();
```

File: src/module.rs [hunk 16]
```diff
@@ -65,6 +65,7 @@ pub const ALL_MODULES: &[&str] = &[
     "username",
     "vcsh",
     "vagrant",
+    "v",
```
Comment (vladimyr) [inline src/module.rs:68]:
```suggestion
    "vlang",
```

File: src/modules/mod.rs [hunk 17]
```diff
@@ -204,6 +206,7 @@ pub fn description(module: &str) -> &'static str {
         "terraform" => "The currently selected terraform workspace and version",
         "time" => "The current local time",
         "username" => "The active user's username",
+        "v" => "The currently installed version of V",
```
Comment (vladimyr) [inline src/modules/mod.rs:209]:
```suggestion
        "vlang" => "The currently installed version of V",
```

**Review by vladimyr — CHANGES_REQUESTED** [review]
Following the precedent established by golang let's use vlang instead. Leaving that aside, good job @Milo123459 :+1:

File: src/configs/v.rs [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+impl<'a> Default for VConfig<'a> {
+    fn default() -> Self {
+        VConfig {
+            format: "via [$symbol($version )]($style)",
+            symbol: "V ",
+            style: "blue bold",
+            disabled: false,
+            detect_extensions: vec!["v"],
+            detect_files: vec![],
```
Comment (vladimyr) [inline src/configs/v.rs:25]:
```suggestion
            detect_files: vec!["v.mod"],
```

File: docs/config/README.md [hunk 19]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
...
+- The current directory contains a file with the extension of .v
+
+### Options
+
+| Option              | Default                                           | Description                                     |
+| ------------------- | ------------------------------------------------- | ----------------------------------------------- |
+| `format`            | `"via [$symbol($version )]($style)"`              | The format for the module.                      |
+| `symbol`            | `"V "`                                           | A format string representing the symbol of V |
+| `detect_extensions` | `["v"]`                                        | Which extensions should trigger this module.    |
+| `detect_files`      | `[]` | Which filenames should trigger this module.     |
```
Comment (vladimyr) [inline docs/config/README.md:2744]:
```suggestion
| `detect_files`      | `["v.mod"]` | Which filenames should trigger this module. |
```

File: docs/config/README.md [hunk 20]
```diff
@@ -2727,6 +2728,40 @@ By default the module will be shown if any of the following conditions are met:
 format = "via [⍱ $version](bold white) "
 ```
 
+## V
+
+The `v` module shows you your currently installed version of V.
+By default the module will be shown if any of the following conditions are met:
```
Comment (vladimyr) [inline docs/config/README.md:2734]:
```suggestion
By default the module will be shown if any of the following conditions are met:

- The current directory contains a `v.mod` file
```

**Review by vladimyr — CHANGES_REQUESTED** [review]
V uses `v.mod` files following Go's `go.mod` approach. @Milo123459 do a quick investigation to find out whether there are additional package management related files with `v.sum` being a likely candidate.

File: docs/config/README.md (hunk 2 above)
Comment (Milo123459) [reply docs/config/README.md:2747]: Hahaha, my bad! 🙃

File: docs/config/README.md (hunk 4 above)
Comment (Milo123459) [reply docs/config/README.md:2762]: I meant blue bold, let me fix it.

File: src/configs/starship_root.rs (hunk 5 above)
Comment (Milo123459) [reply src/configs/starship_root.rs:59]: Okay, let me change that.

**Comment by Milo123459** [comment]
Question, does detect_extensions have to always have that, but also have all the required files?
For example, would:
hello.v without v.mod display? 
Or would it need hello.v & v.mod to display?

**Comment by Milo123459** [comment]
May as well add a version system for V.

**Comment by vladimyr** [comment]
> Question, does detect_extensions have to always have that, but also have all the required files?
> For example, would:
> hello.v without v.mod display? 
> Or would it need hello.v & v.mod to display?

If any of the conditions are met module will get displayed so it is OR rather than AND.

**Comment by vladimyr** [comment]
> May as well add a version system for V. 

Now I lost you. What do you mean by _version system_?

**Comment by Milo123459** [comment]
> > May as well add a version system for V.
> 
> Now I lost you. What do you mean by _version system_?

I updated the package module to display the V version if v.mod is present.

File: docs/config/README.md [hunk 21]
```diff
@@ -2727,6 +2729,41 @@ By default the module will be shown if any of the following conditions are met:
 format = "via [⍱ $version](bold white) "
 ```
 
+## VLang
+
+The `vlang` module shows you your currently installed version of V.
+By default the module will be shown if any of the following conditions are met:
+- The current directory contains a file with the extension of `.v`
```
Comment (vladimyr) [inline docs/config/README.md:2736]:
```suggestion
- The current directory contains a file with `.v` extension
```

File: src/modules/package.rs [hunk 22]
```diff
@@ -46,6 +46,18 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
...
 
+fn extract_vlang_version(file_contents: &str) -> Option<String> {
+    let mut splitted = file_contents.split('\n');
+    let found = splitted.find(|line| line.to_lowercase().contains("version:"))?;
+    let raw_version = found.split("version:").nth(1)?;
+    let version: String = raw_version
+        .chars()
+        .skip(2)
+        .take(raw_version.chars().count() - 3)
+        .collect();
```
Comment (vladimyr) [inline src/modules/package.rs:57]: Replace this with a regex.

**Review by vladimyr — CHANGES_REQUESTED** [review]

**Review by vladimyr — APPROVED** [review]
LGTM 🚀

File: src/configs/mod.rs [hunk 23]
```diff
@@ -124,6 +125,7 @@ pub struct FullConfig<'a> {
     terraform: terraform::TerraformConfig<'a>,
     time: time::TimeConfig<'a>,
     username: username::UsernameConfig<'a>,
+    v: vlang::VLangConfig<'a>,
```
Comment (davidkna) [inline src/configs/mod.rs:128]: I think this also needs to be `vlang`?

```suggestion
    vlang: vlang::VLangConfig<'a>,
```

Comment (vladimyr) [reply src/configs/mod.rs]: Hmm, I was trying to follow golang's suite where the module is called golang and config is called go.

**Comment by vladimyr** [comment]
@Milo123459 Please apply this patch https://github.com/vladimyr/starship/commit/fc96de5.patch

**Review by vladimyr — CHANGES_REQUESTED** [review]
@Milo123459 Please apply this patch https://github.com/vladimyr/starship/commit/8dcd68f.patch

**Comment by Milo123459** [comment]
Those changes are already in place?

**Comment by Milo123459** [comment]
So I'm changing the vlang config to v? I don't know why but sure

**Comment by vladimyr** [comment]
> So I'm changing the vlang config to v? I don't know why but sure

Honestly, this is not my invention 🙃 There is already a Go module that sets precedent by having `src/configs/go.rs` and `src/modules/golang.rs`. And as much as it may be the wrong one I'll always try to keep (naming) consistency. That being said I have zero objections to a new PR that changes both golang, vlang, you name it to use more natural filenames but it has to be consistent across the board. 😉

**Review by vladimyr — APPROVED** [review]
@Milo123459 Once again thanks for your patience and hard work!

LGTM 🚀

File: src/configs/mod.rs (hunk 23 above)
Comment (Milo123459) [reply src/configs/mod.rs]: Ah, makes sense.

**Comment by Milo123459** [comment]
Anything else need to change?

**Comment by andytom** [comment]
Thank you for your contribution @Milo123459 and thank you for reviewing @vladimyr and @davidkna

**Pull request merged**